### PR TITLE
ARROW-10426: [C++] Allow writing large strings to Parquet

### PR DESCRIPTION
**Pull request by pitrou**
Large strings are still read back as regular strings.

### CODE REVIEW

**Comment by pitrou** [comment]
Perhaps I should add tests on the C++ side.

**Comment by pitrou** [comment]
@kou Is it normal that the MinGW tests enable Python?

**Comment by kou** [comment]
Yes. It's normal.

**Comment by pitrou** [comment]
@kou Well, the Python tests sometimes seem to time out on MinGW...

**Comment by kou** [comment]
Umm, they may have a problem in finalization...

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 1]
```diff
@@ -1135,14 +1172,50 @@ TEST_F(TestStringParquetIO, EmptyStringColumnRequiredWrite) {
...
+
+TEST_F(TestLargeStringParquetIO, Basics) {
+  const char* json = R"(["foo", "", null, "bar"])";
+
+  const auto large_type = ::arrow::large_utf8();
+  const auto narrow_type = ::arrow::utf8();
+  const auto large_array = ::arrow::ArrayFromJSON(large_type, json);
+  const auto narrow_array = ::arrow::ArrayFromJSON(narrow_type, json);
+
+  this->RoundTripSingleColumn(large_array, narrow_array,
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:1211]: is this intended to be two different arrays?  same as above?  maybe add a comment on what you are testing here? (lack of schema to read back?)

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -127,6 +129,21 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
   }
 
  protected:
+  template <typename ArrayType>
+  void PutBinaryArray(const ArrayType& array) {
+    const int64_t total_bytes =
+        array.value_offset(array.length()) - array.value_offset(0);
+    PARQUET_THROW_NOT_OK(sink_.Reserve(total_bytes + array.length() * sizeof(uint32_t)));
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:136]: is a check for int32 overflow needed here here?

File: cpp/src/parquet/encoding.cc [hunk 3]
```diff
@@ -592,6 +597,29 @@ class DictEncoderImpl : public EncoderImpl, virtual public DictEncoder<DType> {
   /// Indices that have not yet be written out by WriteIndices().
   ArrowPoolVector<int32_t> buffered_indices_;
 
+  template <typename ArrayType>
+  void PutBinaryArray(const ArrayType& array) {
+    ::arrow::VisitArrayDataInline<typename ArrayType::TypeClass>(
+        *array.data(),
+        [&](::arrow::util::string_view view) {
+          // XXX check length
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:605]: implement the TODO?

File: cpp/src/parquet/encoding.cc [hunk 4]
```diff
@@ -127,6 +129,21 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
   }
 
  protected:
+  template <typename ArrayType>
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:132]: i'm not very familiar with this code, could you let me know what these changes are intended to do?

File: cpp/src/parquet/statistics.cc [hunk 5]
```diff
@@ -691,9 +634,60 @@ void TypedStatisticsImpl<ByteArrayType>::PlainDecode(const std::string& src,
...
 
+}  // namespace
+
 // ----------------------------------------------------------------------
 // Public factory functions
 
+std::shared_ptr<Comparator> Comparator::Make(Type::type physical_type,
+                                             SortOrder::type sort_order,
+                                             int type_length) {
+  if (SortOrder::SIGNED == sort_order) {
```
Comment (emkornfield) [inline cpp/src/parquet/statistics.cc:645]: is this just moving code?

**Review by emkornfield — CHANGES_REQUESTED** [review]
it would be nice to implement the safety checks for max length strings if possible.  Otherwise, most of the comments are questions to help be better understand some of the refactoring.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:1211]: I'll add a comment indeed.

File: cpp/src/parquet/encoding.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:132]: `Put(const ::arrow::Array&)` was hardwired for narrow binary arrays, we need to handle both narrow and large arrays, hence this small refactor.

File: cpp/src/parquet/encoding.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:136]: I don't think so (these are `int64_t` calculations AFAICT). However, I should fix the XXX below :-)

File: cpp/src/parquet/encoding.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:605]: Will do :-)

File: cpp/src/parquet/statistics.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/parquet/statistics.cc:645]: Yes, so as to put internal implementation details in the anonymous namespace.

**Comment by pitrou** [comment]
@emkornfield Any other concern?

File: cpp/src/parquet/encoding.cc [hunk 6]
```diff
@@ -127,6 +130,25 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
...
+  void PutBinaryArray(const ArrayType& array) {
+    const int64_t total_bytes =
+        array.value_offset(array.length()) - array.value_offset(0);
+    PARQUET_THROW_NOT_OK(sink_.Reserve(total_bytes + array.length() * sizeof(uint32_t)));
+
+    constexpr size_t kMaxByteArraySize = std::numeric_limits<uint32_t>::max();
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:139]: its not clear from the spec, but I think this expected to be signed:

[Java](https://github.com/apache/parquet-mr/blob/7469e87dc63c0f672be7175946dc7af1219ec732/parquet-column/src/main/java/org/apache/parquet/column/values/plain/BinaryPlainValuesReader.java#L39) appears to use a int (signed 32 bit integer)

**Comment by emkornfield** [comment]
One concern with the length check, otherwise LGTM.

File: cpp/src/parquet/encoding.cc (hunk 6 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: I've changed the code to check for the signed maximum. It will be safer in any case.

**Pull request closed without merging**